go-controller/pkg/libovsdbops/db_index.go

```diff
@@ -0,0 +1,259 @@
+// - OwnerController
+// - OwnerObjectType
+// - PrimaryIDKey (constructed by DbIndex.GetUniqueID)
+// SecondaryIndexKey key is included if it is returned by dbIndex.ObjectIDs.GetExternalIDs().
+// Object-specific ids (except SecondaryIndexKey) are returned based on DbIndex.ObjectIDs.GetExternalIDs(),
+// with keys prefixed with "<DbIndex.OwnerObjectType>.<DbIndex.OwnerController>.k8s.ovn.org/".
+// Keys with empty string values will be omitted as unset, e.g. for predicate search.
+func (dbIndex *DbIndex[T]) GetExternalIDs() map[string]string {
+	externalIDs := dbIndex.getExternalIDsWithoutPrimary()
+	externalIDs[PrimaryIDKey] = dbIndex.GetUniqueID()
+	return externalIDs
+}
+
+// GetUniqueID returns unique string that identifies object with given DbIndex.
+// Different DbIndexes should always have different ids, equal DbIndexes will return the same string.
+// GetUniqueID uses a value returned by DbIndex.ObjectIDs.GetUniqueIDs, prefixed by
+// DbIndex.ownerKey to make it unique withing the whole db.
+func (dbIndex *DbIndex[T]) GetUniqueID() string {
+	return dbIndex.ownerKey + "/" + dbIndex.ObjectIDs.GetUniqueID()
+}
+
+func (dbIndex *DbIndex[T]) getExternalIDsWithoutPrimary() map[string]string {
+	externalIDs := map[string]string{
+		OwnerController: dbIndex.OwnerController,
+		OwnerObjectType: dbIndex.OwnerObjectType,
+	}
+	for id, value := range dbIndex.ObjectIDs.GetExternalIDs() {
+		if value == "" {
+			// omit empty values
+			continue
+		}
+		if id == SecondaryIndexKey {
+			externalIDs[SecondaryIndexKey] = value
+		} else {
+			externalIDs[dbIndex.ownerKey+"/"+id] = value
+		}
+	}
+	return externalIDs
+}
+
+// GetDbIndexFromExternalIDs is used to parse object ExternalIDs, it uses objectIDsConstructor that should be
+// able to build DbObjectIDs object based on the keys returned by DbObjectIDs.GetExternalIDs().
+func GetDbIndexFromExternalIDs[T DbObjectIDs](externalIDs map[string]string,
+	objectIDsConstructor func(map[string]string) T) (*DbIndex[T], error) {
+	dbIndex := &DbIndex[T]{
+		OwnerController: externalIDs[OwnerController],
+		OwnerObjectType: externalIDs[OwnerObjectType],
+	}
+	// prevent panic from dbIndex.setOwnerKey
+	if dbIndex.OwnerController == "" || dbIndex.OwnerObjectType == "" {
+		return nil, fmt.Errorf("expected keys %s and %s are not found in the external ids %v",
+			OwnerController, OwnerObjectType, externalIDs)
+	}
+	dbIndex.ownerKey = dbIndex.GetOwnerKey()
+	objectIDs := map[string]string{}
+
+	for id, value := range externalIDs {
+		if id == SecondaryIndexKey {
+			objectIDs[id] = value
+		} else {
+			// owner-prefixed id
+			s := strings.Split(id, "/")
+			if len(s) == 2 && s[0] == dbIndex.ownerKey {
+				objectIDs[s[1]] = value
+			}
+		}
+	}
+	dbIndex.ObjectIDs = objectIDsConstructor(objectIDs)
+	return dbIndex, nil
+}
+
+// usesDbIndex interface should only include types that use new ExternalIDs from DbIndex.
+type usesDbIndex interface {
+	GetExternalIDs() map[string]string
+	*nbdb.AddressSet
```

Comment:
Why do we need `*nbdb.AddressSet` here? How would this look like with more types then?

Reply:
it will be `*nbdb.AddressSet | *nbdb.ACL` and is used as a type constraint for functions that only make sense for db objects that are using new ids

Reply:
Ok, I did not know that you could mix type constraints and methods on the same interface and I was actually incorrectly interpreting this as embedding.

But again, what is the actual use of the type constraint & parameter once we have the interface with the `GetExternalIDs` method? I am just sensing a bit of overuse of generics but I might be missing something.

Reply:
So still asking myself why would we need this as type parameter/type assertion. If used as a regular interface, wouldn't we get a compile time error if a non-implementing is passed?